Various fixes

Chaphasilor opened the pull request:
## Changes

- Lyrics are now always fetched during a download repair, for each track. This is the only way to ensure a full repair/sync actually updates the lyrics
  - fixes multiple reports on Discord
- Updated downloads file structure for human-readable (i.e., external) downloads
  - fixes #1391 (although we still create a "Finamp" directory, I wasn't sure if this is strictly necessary)
  
I might keep adding things here, where I'd like another pair of eyes to take a look.

Chaphasilor committed:
always fetch lyrics during downloads repair

- there's no other way to figure out if have been changed, and a repair should ensure the downloads are consistent with the server

Chaphasilor committed:
improve human-readable file structure

Komodo5197 commented on lib/services/downloads_service_backend.dart:
```diff
@@ -1552,12 +1555,45 @@ class DownloadsSyncService {
+    String fileName;
+    String subDirectory;
+    if (downloadLocation.useHumanReadableNames) {
+      fileName = _filesystemSafe(
```
I'm not sure I like the idea of assigning the item name to images.  Assigning a specific track's name to an image that could be associated with, as an example, every track in the album and the album itself could be a bit misleading, and if the album itself gets selected I think the image would end up outside the album folder?  And we're still going to end up with a bunch of images for things like artists and genres cluttering up the root directory.  It might make sense to just give up on human readable image locations and stick all the images into a dedicated images folder?

Chaphasilor replied:
So my reasoning was to preserve as much metadata in the filenames as possible. The naming should never be incorrect, we just avoid duplicates if all other tracks have the same cover. So I don't really see the issue with doing things this way.  
I also noticed the issue with album covers ending up outside the album directory, but - although annoying - this doesn't seem too problematic.  
One thing I was debating is just throwing all images into an `images` subdirectory, that could also work to keep things tidy.  
Finally, Jellyfin does expose the original file path on the server, so we could use that to exactly re-create the server-side file structure (sanitized of course). That might be an elegant solution that actually would be usable as a (partial) library backup.

Komodo5197 replied:
If we reliably get the full server path, just using that wouldn't be crazy.  Although that introduces the possibility of issues relating to parsing path separators and usable characters across different filesystems.  I do like using the original filename, at the very least.  Although maybe we should add a random id at the end to handle possible collisions due to possibly grouping items which were in separate folders.  I think the previous method also could theoretically generate collisions.  

I think I'd prefer just shoving everything into an images folder.  The reason I don't like the track names is that it implies precision we don't have.  You see an image for track 13 and wonder where all the other track images are.  I don't think having the images in the audio folder structure is actually useful anyway.

Chaphasilor replied:
then I'll move images into their own subdirectory and go back to using UUIDs, and also append a random ID to each original file name 👍

Komodo5197 commented on lib/services/downloads_service_backend.dart:
```diff
@@ -1552,12 +1555,45 @@ class DownloadsSyncService {
+    String subDirectory;
+    if (downloadLocation.useHumanReadableNames) {
+      fileName = _filesystemSafe(
+        "${item.indexNumber != null ? "[${item.indexNumber}] " : ""}${item.artists?.isNotEmpty ?? false ? "${item.artists?.joinNonNull(", ")} - " : ""}${item.name}",
```
Is it really worth putting all the artists into the filenames?  This could potentially make them very long.

Chaphasilor replied:
You're right, that could be an issue. I was trying to have a consistent format, and since Jellyfin doesn't expose a "main" artist I wanted to include all. I'll switch this to use the original file name (if available), or fall back to just the first artist.

Chaphasilor committed:
update base gain for emulated volume normalization (iOS & desktop)

- this is done to increase the default volume of Finamp by default. it gives us less headroom to boost quiet tracks, but should be closer to 100% volume for most other tracks.

Chaphasilor commented:
FYI while playing around with this I noticed that when a downloaded album is loaded on the albums tab, and then gets deleted, it will have a blank cover until the tab is reloaded. Might be related to your changes to image caching, if I remember correctly?

Chaphasilor committed:
use original file name for human-readable downloads

Chaphasilor committed:
make local network address input more robust

- fixes #1328

Chaphasilor committed:
suppress DNS errors & duplicate network errors

- this is an experimental change to see if we can get away with showing less errors, reducing snackbar spam when something goes wrong

Komodo5197 commented on lib/components/global_snackbar.dart:
```diff
@@ -117,5 +159,14 @@ class GlobalSnackbar {
     );
+
+    // When this snackbar fully closes (timeout or swipe), clear the key if it matches, allowing future identical errors.
+    if (isNetworkError && dedupKey != null) {
```
We don't need the isNetweorkError check here.

Komodo5197 commented on lib/components/NetworkSettingsScreen/prefer_local_network_address_selector.dart:
```diff
@@ -10,16 +10,55 @@ import 'package:get_it/get_it.dart';
   const LocalNetworkAddressSelector({super.key});
 
+  // Externally callable hook to force committing pending edits before actions (e.g., test connection)
+  static Future<void> Function()? commitPendingEdits;
```
I'd prefer this not be global, even if we only expect one widget at once.  I think the way to do this is to give the widget a key on creation, and then call key.currentState.commitIfChanged.

Komodo5197 commented on lib/components/NetworkSettingsScreen/prefer_local_network_address_selector.dart:
```diff
@@ -29,6 +68,9 @@ class _LocalNetworkAddressSelector extends ConsumerState<LocalNetworkAddressSele
     bool featureEnabled = user?.preferLocalNetwork ?? DefaultSettings.preferLocalNetwork;
 
     _controller ??= TextEditingController(text: address);
```
We shoulkd set up _controller and _lastcommittedValue inside of initState.

Komodo5197 commented on lib/components/NetworkSettingsScreen/prefer_local_network_address_selector.dart:
```diff
@@ -39,15 +81,18 @@ class _LocalNetworkAddressSelector extends ConsumerState<LocalNetworkAddressSele
-            }
-            GetIt.instance<FinampUserHelper>().currentUser?.update(newLocalAddress: value);
-            await changeTargetUrl();
+          onTapOutside: (_) {
```
Is all of this needed with the focus listener?  I don't know much about the focus system, but it feels like at least some of this should be redundant.

Chaphasilor replied:
I honestly haven't tested this extensively, I just wanted to make sure it actually saves the value

Chaphasilor committed:
improve URL value change handling

Komodo5197 commented on lib/screens/network_settings_screen.dart:
```diff
@@ -33,14 +33,16 @@ class _NetworkSettingsScreenState extends State<NetworkSettingsScreen> {
               // Ensure any pending edits in the local network address field are committed first
-              if (LocalNetworkAddressSelector.commitPendingEdits != null) {
-                await LocalNetworkAddressSelector.commitPendingEdits!();
+              final widgetState = LocalNetworkAddressSelector.localNetworkAddressKey.currentState;
```
This key doesn't need to be a static global, just create it as a local variable.  I think if you make the _LocalNetworkAddressSelector class public you can set that as the key type and you shouldn't need to cast anything, but I'm not 100% sure of that.

Chaphasilor replied:
thanks for the tip, that worked.

Komodo5197 commented on lib/components/NetworkSettingsScreen/prefer_local_network_address_selector.dart:
```diff
@@ -10,28 +10,34 @@ import 'package:get_it/get_it.dart';
   @override
   void initState() {
     super.initState();
+    FinampUser? user = ref.read(FinampUserHelper.finampCurrentUserProvider).valueOrNull;
```
As mentioned in the other PR, I don't really like using ref.read, and would prefer to just directly access the setting.

Komodo5197 commented:
Regarding the album cover being blank after deleting, this is presumably related to the image caching or providers, but I'm not really sure how it would have happened.  If we re-evaluated the albumImageProvider we should have gotten a correct reference to the network image, and if we didn't we would still have the exact same file image provider, which you would expect to hit the decoded image in the flutter image cache.  I made a basic try to reproduce this but failed.  Had the relevant image been involved in the queue at any point, because the logic there is a bit more complex and has more room for errors.

Chaphasilor commented:
Hmm, I just tried reproducing it just now and it's not happening anymore, no matter how I download, queue, hot restart, refresh, or let images fail to download. It was very consistent for me at some point though, but whatever. If I encounter it again I'll try to record better reproduction steps.

Chaphasilor committed:
add ID to track file, save images into separate dir, simplify global key

Chaphasilor committed:
only create Finamp subdirector for humand-readable locations, format

Chaphasilor commented:
Okay, this should be done now. Images are now stored in an `images` subdirectory using a plain UUID, and original track names are combined with metadata-based directory trees and the first 8 characters of the item ID to get a robust structure.  
I've opted not to use the path as I'm not sure how long it will be around. I've heard that the path was only kept because the Kodi addon used it for direct playback of media files, which is no longer supported, so I'm not too confident basing logic on this (even if there's a fallback).

Feel free to merge if this looks good to you!

Komodo5197 commented:
The comments and code still suggest that _getDownloadPath is used for both tracks and images, but it looks like you broke them up again and images are handled internally, so you should clean that up.  Other than that, it looks good to me.

Chaphasilor committed:
make separation of track and image download locations clearer, regenerate code

Chaphasilor commented:
Ah, seems like I missed a part of the comment when removing the references to images. Should be clearer now.
